engines/sherlock/events.cpp

```diff
@@ -0,0 +1,299 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/scummsys.h"
+#include "common/events.h"
+#include "common/system.h"
+#include "engines/util.h"
+#include "graphics/cursorman.h"
+#include "sherlock/sherlock.h"
+#include "sherlock/events.h"
+
+namespace Sherlock {
+
+Events::Events(SherlockEngine *vm) {
+	_vm = vm;
+	_cursorImages = nullptr;
+	_cursorId = INVALID_CURSOR;
+	_frameCounter = 1;
+	_priorFrameTime = 0;
+	_mouseButtons = 0;
+	_pressed = _released = false;
+	_rightPressed = _rightReleased = false;
+	_oldButtons = _oldRightButton = false;
+}
+
+Events::~Events() {
+	delete _cursorImages;
+}
+
+/**
+ * Load a set of cursors from the specified file
+ */
+void Events::loadCursors(const Common::String &filename) {
+	hideCursor();
+	delete _cursorImages;
+
+	_cursorImages = new ImageFile(filename);
+	_cursorId = INVALID_CURSOR;
+}
+
+/**
+ * Set the cursor to show
+ */
+void Events::setCursor(CursorId cursorId) {
+	if (cursorId == _cursorId)
+		return;
+
+	_cursorId = cursorId;
+
+	// Set the cursor data
+	Graphics::Surface &s = (*_cursorImages)[cursorId];
+
+	setCursor(s);
+}
+
+/**
+ * Set the cursor to show from a passed frame
+ */
+void Events::setCursor(const Graphics::Surface &src) {
+	CursorMan.replaceCursor(src.getPixels(), src.w, src.h, 0, 0, 0xff);
+	showCursor();
+}
+
+/**
+ * Show the mouse cursor
+ */
+void Events::showCursor() {
+	CursorMan.showMouse(true);
+}
+
+/**
+ * Hide the mouse cursor
+ */
+void Events::hideCursor() {
+	CursorMan.showMouse(false);
+}
+
+/**
+ * Returns the cursor
+ */
+CursorId Events::getCursor() const {
+	return _cursorId;
+}
+
+/**
+ * Returns true if the mouse cursor is visible
+ */
+bool Events::isCursorVisible() const {
+	return CursorMan.isVisible();
+}
+
+/**
+ * Move the mouse
+ */
+void Events::moveMouse(const Common::Point &pt) {
+	g_system->warpMouse(pt.x, pt.y);
+}
+
+
+/**
+ * Check for any pending events
+ */
+void Events::pollEvents() {
+	checkForNextFrameCounter();
+
+	Common::Event event;
+	while (g_system->getEventManager()->pollEvent(event)) {
+		// Handle keypress
+		switch (event.type) {
+		case Common::EVENT_QUIT:
+		case Common::EVENT_RTL:
+			return;
+
+		case Common::EVENT_KEYDOWN:
+			// Check for debugger
+			if (event.kbd.keycode == Common::KEYCODE_d && (event.kbd.flags & Common::KBD_CTRL)) {
+				// Attach to the debugger
+				_vm->_debugger->attach();
+				_vm->_debugger->onFrame();
+			} else {
+				_pendingKeys.push(event.kbd);
+			}
+			return;
+		case Common::EVENT_KEYUP:
+			return;
+		case Common::EVENT_LBUTTONDOWN:
+			_mouseButtons |= 1;
```

Comment:
Similarly, to my comment about `_mousePos` below: EventManager already stores this button state. In fact it is using the same bits for left and right mouse buttons. And it features constants for readability. You might want to consider using it instead of duplicating code.